---
title: "Accompanying code for annotating single-cell maps"
output: html_notebook
---

Before beginning this tutorial, make sure you have the following packages installed:

R version | Packages                    |                |
----------|-----------------------------|----------------|-----------------
4.0.2     |SingleCellExperiment_1.11.7* |dplyr_1.0.2*    |ggplot2_3.3.2
         .|Seurat_3.2.1                 |scmap_1.11.0*   |harmony_1.0*
         .|scater_1.17.5*               |celldex_0.99.1* |cerebroApp_1.2.2*
         .|SCINA_1.2.0                  |SingleR_1.3.8*  |msigdb_0.2.0*
"*" = packages must be installed by running `BiocManager::install("package")` instead of `install.packages("package")`

# 1. Reference-based automatic annotation

### Create the Reference
The first step in performing reference-based annotation is to select an annotated dataset to use as the reference. Here we will use one of the references created by the authors of SingleR and show how it can be used with another tool.

Other reference datasets can be found in GEO (https://www.ncbi.nlm.nih.gov/geo/) or at a link provided by the authors of the reference dataset. However, to use a dataset as a reference you will need both the single-cell RNA sequencing data and the cell-type annotations. GEO does not require authors to provide the cell-type annotations of their data, so you may need to contact the authors directly to to get the annotations for some datasets.

```{r, echo = TRUE, results = 'hide'}
# Set a random seed to ensure result reproducibility
set.seed(9742)
# Download singleR reference data for immune cells and save it as the variable "ref"
# The variable is a class called "Summarized Experiment"
library(celldex)
ref <- DatabaseImmuneCellExpressionData()
```

Next we need to reformat the data to ensure it is compatible with the tool we are using. We will be demonstrating **scmap**, which uses data formatted as a 'SingleCellExperiment object', and assumes by default that gene names are found in a column named 'feature_symbol' while the cell-type labels are in a column named 'cell_type1'. In addition, scmap requires that you normalize and long-transform the reference data; this has already been done for the SingleR reference data so we skip those steps here.

```{r}
# Assign cell-type labels in a column named "cell_type1"
colData(ref)$cell_type1 <- colData(ref)$label.fine
# Assign gene names in a column called "feature_symbol"
rowData(ref)$feature_symbol <- rownames(ref)

# Convert the data into a SingleCellExperiment object
require(SingleCellExperiment)
ref_sce <- SingleCellExperiment(assays=list(logcounts=Matrix::Matrix(assays(ref)$logcounts)), 
			colData=colData(ref), rowData=rowData(ref))
```

Our reference data is ready to me used now. So lets process this data to build 
the index we will use to map our un-labelled data to. First, we select genes to use:

```{r}
require(scmap)
# Create scmap-cluster reference by first selecting the most informative features
ref_sce <- selectFeatures(ref_sce, suppress_plot=FALSE)

# Inspect the first 50 genes selected by scmap
rownames(ref_sce)[which(rowData(ref_sce)$scmap_features)][1:50]
```

Now we can see the genes that scmap has chosen to use. If there are key markers missing
we can make sure they are included like this:

```{r}
# Create a list of key markers
my_key_markers = c("TRAC", "TRBC1", "TRBC2", "TRDC", "TRGC1", "TRGC2", "IGKC")
# Ensure markers are in the list of features used by scmap
rowData(ref_sce)$scmap_features[rownames(ref_sce) %in% my_key_markers] <- TRUE
```

And we can remove genes that we think might be technical artefacts, such as mitochondria RNAs, like this:

```{r}
# Create a list of mitocondrial genes from the dataset (genes that begin with "MT")
mt_genes <- rownames(ref_sce)[grep("^MT-", rownames(ref_sce))]
# Remove these genes from the features used by scmap
rowData(ref_sce)$scmap_features[rownames(ref_sce) %in% mt_genes] <- FALSE

# Extract the features and assign them to a new variable, "scmap_feature_genes"
scmap_feature_genes <- rownames(ref_sce)[which(rowData(ref_sce)$scmap_features)]
```

Now we build the reference profiles used in **scmap-cluster**, for **cluster-based cell-type annotation**. These profiles can be accessed and plotted from inside the SingleCellExperiment object as follows:

```{r}
# Create reference profiles;
# Once reference profiles are generated the original data are 
# not needed for scmap-cluster
ref_sce <- scmap::indexCluster(ref_sce)
# Visualize interesting features as a heatmap
heatmap(as.matrix(metadata(ref_sce)$scmap_cluster_index))
# Store expression information as a variable
scmap_cluster_reference <- metadata(ref_sce)$scmap_cluster_index
```

From here on out scmap only needs this set of reference profiles, so if working with a very large reference one could save this index separately and load it directly to use when annotating new datasets. But since that is not the case here, we will simply save this index to a variable for now. 

We will also demonstrate **scmap-cell** to **annotate individual cells** of our dataset, so we will create that index as well. As before one would first normalize and log-transform the reference data, and select genes to use. As we have already done that, we need only run the command to build the scmap-cell index. There are two parameters we can set: M and k, increasing M and k will give more accurate mapping but increase the size of the index, and the time needed to map cells. Here we use the defaults:

```{r}
# Create scmap-cell reference
ref_sce <- scmap::indexCell(ref_sce)
scmap_cell_reference <- metadata(ref_sce)$scmap_cell_index
scmap_cell_metadata <- colData(ref_sce)
```

scmap-cell assigns cells in one dataset to their "nearest neighbours" in the reference dataset. In this case, the "nearest neighbours" are the cells in the reference dataset most similar to the cells in the query dataset.

One can use any rule they like to transfer information, such as cell-type or pseudotime, from these nearest neighbours to the query data. Thus we need to store the associated metadata for the reference as well. Now we don't need to use our original reference dataset anymore.

### Assign cells from the query dataset to the reference.

The query dataset we will be using is provided by 10X genomics. The data can be retrieved by writing the following on the command line.

`wget https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz`
`tar -xvzf pbmc3k_filtered_gene_bc_matrices.tar.gz`
The data can now be found in your directory `./filtered_gene_bc_matrices/hg19/`

An alternative method is to copy and paste the same link into your browser, and the data will automatically download and can subsequently be extracted.

Now we need to load our unlabeled dataset into R. Normal preprocessing including QC filtering, normalizing and log-transforming the data must be done prior to annotating. In addition, scmap is based on the **SingleCellExperiment object**, so if our data is stored as a Seurat object we must convert it to SingleCellExperiment as shown below.


```{r}
require(Seurat)
# This portion of the tutorial is assuming the raw 10X data is in the following folder in
# your directory:
data <- Read10X("filtered_gene_bc_matrices/hg19/")
# Make SingleCellExperiment from the raw matrix
query_sce <- SingleCellExperiment(assays=list(counts=data))

# Make SingleCellExperiment from Seurat object
query_seur <- CreateSeuratObject(data)
query_sce <- as.SingleCellExperiment(query_seur)

# normalize the data using the scater package
query_sce <- scater::logNormCounts(query_sce)

# add feature_symbol column
rowData(query_sce)$feature_symbol <- rownames(query_sce)
```

Now you should have an entry in `assays(my_sce)` called `logcounts` with the log-normalized matrix. We are now ready to annotate our data with **scmap-cluster**. Let's start with scmap-cluster:

```{r}
# Run scmapCluster
scmap_cluster_res <- scmapCluster(projection=query_sce, 
				index_list=list(immune1 = scmap_cluster_reference), 
				threshold=0.1)

# plot the results of our annotation
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cluster_res$combined_labs), las=2)

# Store this annotation information within the query object
colData(query_sce)$scmap_cluster <- scmap_cluster_res$combined_labs

# Make a UMAP of the cells, labeled with the cell-type annotations from scmapCluster
query_sce <- scater::runUMAP(query_sce)
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="scmap_cluster")
```

Alternatively we could use **scmap-cell**, to find the 10 nearest neighbours to each cell (i.e. the 10 most similar cells to each query cell), then pick the annotation that is most common among the neighbours like this:

```{r}
# Determine the 10 "nearest neighbours" from the reference dataset for each cell in the query dataset
# using scmapCell
nearest_neighbours <- scmapCell(projection=query_sce, 
	index_list = list(immune1 = scmap_cell_reference), 
	w=10)

# Get metadata for the neighbours of each cell.
mode_label <- function(neighbours, metadata=scmap_cell_metadata$cell_type1) {
	freq <- table(metadata[neighbours])
	label <- names(freq)[which(freq == max(freq))]
	if (length(label) > 1) {return("ambiguous")}
	return(label)
}

scmap_cell_labs <- apply(nearest_neighbours$immune1$cells, 2, mode_label)

#add them to the query object
colData(query_sce)$scmap_cell <- scmap_cell_labs

# plot the results
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cell_labs), las=2)

# Make a UMAP and add the cell-type annotations
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="scmap_cell")
```

Another option compatible with the SingleCellExperiment Object is **SingleR**. As before, we need a reference and a query dataset. In the case of SingleR, we need the entirety of the reference dataset, rather than generating a compressed
reference index as we did with scmap. In addition, running just this small example demonstrates the difference in run time between the methods. 

```{r}
require(SingleR)
# Run SingleR on the query data and the reference
predictions <- SingleR(test=query_sce, ref=ref, labels=ref$label.fine)
colData(query_sce)$singleR <- predictions$pruned.labels

# plot the results
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cell_labs), las=2)

# Make a UMAP and add the cell-type annotations
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="singleR")
```

### Integration as a form of annotation

Another option is to integrate our query data with our reference data. Then we simply transfer the labels from the annotated reference to the neighbouringquery cells in the integrated dataset. Clustering the integrated data is a common approach to transferring labels. We demonstrate how this could be done with **Harmony** below. But the approach would be the same for any
integration tool.

Note: the SingleR reference is not single cells, but averages across many cells. Thus we convert and downsample the reference to a single cell object for demonstration purposes.For a real experiment, one would use the original single cells as the reference when integrating datasets.


```{r}
set.seed(2891)
# Convert reference and query datasets to Seurat Objects

# Add a "counts" slot to the reference SingleCellExperiment object so we can convert it to a Seurat Object
assays(ref_sce)[["counts"]] <- round(2^assays(ref_sce)[["logcounts"]]) -1
colnames(ref_sce) <- paste("cell", 1:ncol(ref_sce))

# Subset both objects so both the reference and query datasets have the same genes
ref_seur <- as.Seurat(ref_sce[rownames(ref_sce) %in% rownames(query_sce),])
ref_seur@active.ident <- factor(rep("reference", ncol(ref_seur)))

query_seur <- as.Seurat(query_sce[rownames(query_seur) %in% rownames(ref_sce),])
query_seur@active.ident <- factor(rep("query", ncol(query_seur)))

# Downsample the reference to be similar to query in terms of total UMIs
totalUMI <- median(query_seur@meta.data$nCount_RNA)
ref_seur@assays$RNA@counts <- SampleUMI(ref_seur@assays$RNA@counts, max.umi=totalUMI, upsample=FALSE)

# Merge the datasets together into a single Seurat object
merged_seur <- merge(ref_seur, query_seur)
merged_seur@meta.data$source <- merged_seur@active.ident

# Normalize the data
require(harmony)
merged_seur <- NormalizeData(merged_seur)

# For consistency we use the scmap feature genes
VariableFeatures(merged_seur) <- scmap_feature_genes

require(ggplot2)

# Scale the data and run dimensionality reduction
merged_seur <- ScaleData(merged_seur)
merged_seur <- RunPCA(merged_seur)
merged_seur <- RunUMAP(merged_seur, dims=1:15)
DimPlot(merged_seur, reduction="umap")+ggtitle("Before Integration")
# Run Harmony to remove batch effects
merged_seur <- RunHarmony(merged_seur, "source", dims.use=1:15)
merged_seur <- RunUMAP(merged_seur, dims=1:15, reduction="harmony")
# Plot the data
DimPlot(merged_seur, reduction="umap")+ggtitle("After Integration")
```

Now that the data is integrated we will cluster the data and look at the annotations of the reference cells present in each cluster. As with all clustering, this may require manual tuning of the resolution parameters to get the best labels.

```{r}
# Cluster the integrated dataset
merged_seur <- FindNeighbors(merged_seur, reduction="harmony", dims=1:15)
merged_seur <- FindClusters(merged_seur, resolution=0.5)
# Plot the data
DimPlot(merged_seur, reduction="umap")+ggtitle("After Integration")
# Create a table of cluster labels based on integrated data
table(merged_seur@meta.data$label.fine, 
		merged_seur@active.ident)
```


Here we have a table of the reference annotations (across rows) per cluster (across columns). 
We can manually label the clusters based on this table or we could create a rule to algorithmically label the clusters based on this table. Since there is only 10 clusters, we assign the labels manually.

```{r}
cluster_labs <- c("0"="ambiguous", 
	"1"="NK cells", 
	"2"="Monocytes, CD14+", 
	"3"="B cells, naive",
	"4"="T cells, CD4+, naive TREG",
	"5"="T cells, CD8+, naive",
	"6"="Monocytes, CD16+",
	"7"="T cells, CD4+, memory TREG ",
	"8"="T cells, CD4+, naive, stimulated ",
	"9"="T cells, CD8+, naive, stimulated")

# Assign cluster label to the associated query cells
merged_seur@meta.data$annotation <- cluster_labs[merged_seur@meta.data$RNA_snn_res.0.5]

# Add the results to the SingleCellExperiment Object and plot
query_sce$Harmony_lab <- merged_seur@meta.data$annotation[merged_seur@meta.data$source =="query"]
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="Harmony_lab")
```

# 2. Refining / Consensus annotations

Once we have run several tools, we can use the consensus of the labels to get a 
more robust annotation. In this case we will simply use the most common label across tools
to assign the final automatically annotated label.

```{r}
annotation_columns <- c("scmap_cluster", "scmap_cell", "singleR", "Harmony_lab")

#Optional check how consistent the labelling was.
#head(colData(query_sce)[,annotation_columns])

get_consensus_label <- function(labels){
	labels <- labels[labels != "ambiguous"]
	if (length(labels) == 0) {return("ambiguous")}
	freq <- table(labels)
	label <- names(freq)[which(freq == max(freq))]
	if (length(label) > 1) {return("ambiguous")}
	return(label)
}

colData(query_sce)$consensus_lab <- apply(colData(query_sce)[,annotation_columns], 1, get_consensus_label)
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="consensus_lab")
```

# 3. Marker-based automatic annotation

An alternative way for annotation of your query scRNAseq dataset is to utilize Marker-based annotation tools. **SCINA** is a **semi-supervised annotation tool** that takes in the signature genes and expression matrix and predicts the potential labels based on the prior knowledge of the cell-type-specific markers. List of markers is usually provided in the gmt format. The PBMC gene set used bellow have been gathered by [Diaz-Mejia JJ et al.](https://zenodo.org/record/3369934#.X2PWty2z1QI)

The data can be gathered directly by downloading and extracting the pbmc_22_10x.tar.bz2 file, or entering the following on the command line:
`wget https://zenodo.org/record/3369934/files/pbmc_22_10x.tar.bz2?download=1`
`tar -xvjf pbmc_22_10x.tar.bz2`
The extracted data will by located in the following file:
`./MY_PAPER/SUPPLEMENTARY_DATA/pbmc_22_10x/pbmc_22_10x_cell_type_signature_gene_sets.gmt`

The results from this annotation tool are not used in the above step to find consensus annotations because the lists of marker genes are not consistent with the cell types identified in the reference dataset. This is because these data come from different sources, and would not have been characterizing the exact same set of cells. If you wish for marker-based and reference-based annotation methods to be combined in the above step of automatically determining consensus annotations, you would have to make sure all of the identified cell subtypes are the same and that they are spelt the exact same way in order for R to recognize the names as identical.


```{r}
require(SCINA)
require(msigdb)

# Import the marker genes as a GMT file and store as a variable
markers <- read.gmt('./MY_PAPER/SUPPLEMENTARY_DATA/pbmc_22_10x/pbmc_22_10x_cell_type_signature_gene_sets.gmt')
# Convert the expression data from Seurat object into a matrix data structure
exprMatrix <- as.matrix(GetAssayData(query_seur))
# Run SCINA on the query data using the marker genes to identify cell types
# Specifying rm_overlap = FALSE allows the same marker gene to specify multiple cell types which
# may be useful if identifying cell subtypes or other similar types of cells
# Specifying allow_unknown = TRUE allows cells to be labeled as "unknown" instead of being
# assigned a low-confident label
predictions.scina = SCINA(exp = exprMatrix, signatures = markers$genesets,
                          rm_overlap = FALSE, allow_unknown = TRUE)
# Add SCINA annotation information to each cell in Seurat object
colData(query_sce)$SCINA <- predictions.scina$cell_labels

# Make a UMAP and add the SCINA cell-type annotations
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="SCINA")

```

# 4. Manual annotation

### Retrieving marker genes

If you do not have an extensive list of markers per cell type, or a good quality reference dataset, it is useful to extract the top marker genes from each cluster of your query data. We can easily do this in **Seurat**, with the data formatted as a ***Seurat object** (which we created earlier and stored as the variable `query_seur`). First, the data must be normalized and scaled, and the variable genes between cells must be determined.

```{r}
require(Seurat)
query_seur <- NormalizeData(query_seur) # Normalize the data
query_seur <- FindVariableFeatures(query_seur) # Determine the variable features of the dataset
query_seur <- ScaleData(query_seur) # Scale the data based on the variable features
```

Next, different types of dimensionality reduction must be performed on the data so that the cells can be grouped together in 2D space.

```{r}
query_seur <- RunPCA(query_seur)
query_seur <- RunTSNE(query_seur)
query_seur <- RunUMAP(query_seur, dims = 1:50)
```

From this object, we can cluster the data at a chosen resolution that can be modified later on if desired.

```{r}
# Determine the "nearest neighbours" of each cell
query_seur <- FindNeighbors(query_seur, dims = 1:50)
# Cluster the cells
query_seur <- FindClusters(query_seur, resolution = 0.5)
```

Before extracting the marker genes, let's visualize our data on a UMAP.

```{r}
DimPlot(query_seur, reduction = "umap")
```

Now let's extract the top marker genes, and see which ones correspond with each cluster. This can be done using the FindAllMarkers function within Seurat.

```{r, echo = TRUE, results = 'hide'}
markers_seur <- FindAllMarkers(query_seur, only.pos = TRUE)
```

```{r}
# Markers are now stored and can be viewed in the following table. They are ordered from lowest to highest p-value percluster:
markers_seur
```

Marker genes are commonly viewed as a heatmap, which can be created as such:

```{r}
require(dplyr)
# Retrieve the top 10 marker genes per cluster
top10 <- markers_seur %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
# Create the heatmap
DoHeatmap(query_seur, features = top10$gene) + NoLegend()
```

### Pathway analysis

Pathway analysis can also be done for each cluster to determine significantly up- and downregulated pathways based on known gene function. An easy way to do this is by feeding our current Seurat object into **cerebroApp**. Since pathway analysis is based on the differential expression of genes, marker genes must be fetched again through cerebro before performing simple pathway analysis.

```{r, echo = TRUE, results = 'hide'}
require(cerebroApp)

# First get marker genes through cerebro
query_seur <- getMarkerGenes(query_seur,
                      assay = "RNA",
                      organism = "hg",
                      column_sample = "orig.ident",
                      column_cluster = "seurat_clusters")

# Get enriched pathways through cerebro
query_seur <- getEnrichedPathways(query_seur,
                    column_sample = "orig.ident",
                    column_cluster = "seurat_clusters",
                    databases = c("GO_Biological_Process_2018",
                                  "GO_Cellular_Component_2018",
                                  "GO_Molecular_Function_2018",
                                  "KEGG_2016",
                                  "WikiPathways_2016",
                                  "Reactome_2016",
                                  "Panther_2016",
                                  "Human_Gene_Atlas",
                                  "Mouse_Gene_Atlas"),
                    adj_p_cutoff = 0.05,
                    max_terms = 100,
                    URL_API = "http://amp.pharm.mssm.edu/Enrichr/enrich")
```

```{r}
# Enriched pathways are stored in the following location:
query_seur@misc$enriched_pathways
```

Combining top marker genes with functional pathway information should be fairly indicative of cell type depending on your selected clustering resolution. You can easily rerun the analyses after modifying the resolution to get a better idea of subtypes (increase resolution) or identify more general cell types (decrease resolution).
